api/client/cp: provide progress indicator for docker cp

yangdongsheng opened the pull request:
closes #16532

This commit make docker cp output the progress in copying,
something like docker pull.

Signed-off-by: Dongsheng Yang yangds.fnst@cn.fujitsu.com

yangdongsheng commented:
This PR tries to address the feature request in https://github.com/docker/docker/issues/16532

@ErikGrimes

cpuguy83 commented:
Note, this should probably only display progress if `cli.isTerminalOut` is true, and also note that you should use `cli.out` instead of `os.Stdout`

Other than that, +1

yangdongsheng commented:
Thanx @cpuguy83 , This branch is updated, please help to review. :)

yangdongsheng commented:
ping @jlhawn @cpuguy83  Please help to review.

cpuguy83 commented:
@yangdongsheng maintainers are currently preparing for a release, so this may take some time to review.

yangdongsheng commented:
@cpuguy83 Okey, got it. Thanx for your reply.

yangdongsheng commented:
ping @jlhawn @cpuguy83 @thaJeztah Pleaes help to review.

coolljt0725 commented on api/client/cp.go:
```diff
@@ -210,10 +212,28 @@ func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) (
+			In:             response.body,
+			Out:            cli.out,
+			Formatter:      sf,
+			Size:           stat.Size,
```
If copy a dir, the `stat.Size` is wrong, it's not the totally size of the dir.
Maybe you can use https://github.com/docker/docker/blob/master/pkg/directory/directory_unix.go#L12 to get the total size of a directory if `stat.Mode.IsDir()` is `true`

yangdongsheng replied:
@coolljt0725  Oh, yes. Thanx for your comment. Will update it soon. :)

yangdongsheng commented:
@coolljt0725 Updated.
Result:
cp_test is a dir containing a file of 100M.
```
# ./bundles/1.9.0-dev/binary/docker cp cp_test busybox:/
Copying [==================================================>] 104.9 MB/104.9 MB
# ./bundles/1.9.0-dev/binary/docker cp busybox:/cp_test .
Copying [==================================================>] 104.9 MB/104.9 MB
```

LK4D4 commented:
@yangdongsheng seems like your PR doesn't compile for darwin.

LK4D4 commented:
Design sounds ok for me

yangdongsheng commented:
@LK4D4 Sorry for the late reply. I rebased my branch. But I found an error in CI, not sure it's related with my patch.
```
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/include/python2.7 -c coverage/tracer.c -o build/temp.linux-x86_64-2.7/coverage/tracer.o
    coverage/tracer.c:3:20: fatal error: Python.h: No such file or directory
     #include "Python.h"
```

LK4D4 commented:
@yangdongsheng You need to add `darwin` build tag to `pkg/directory/directory_unix.go`

yangdongsheng commented:
@LK4D4 Thanx a lot. updated, then there is only a error in windows building. Others are all GREEN now.

icecrime commented:
Thanks, moving to code review!

LK4D4 commented:
@yangdongsheng This breaks `-` usage:
```
docker cp 699e6d:/etc/passwd -
passwd0100664000000000000000000000056312532674766010521 0ustar0000000000000000root:x:0:0:root:/root:/bin/sh
daemon:x:1:1:daemon:/usr/sbin:/bin/false
bin:x:2:2:bin:/bin:/bin/false
sys:x:3:3:sys:/dev:/bin/false
sync:x:4:100:sync:/bin:/bin/sync
mail:x:8:8:mail:/var/spool/mail:/bin/false
www-data:x:33:33:www-data:/var/www:/bin/false
operator:x:37:37:Operator:/var:/bin/false
ftp:x:83:83:ftp:/home/ftp:/bin/false
nobody:x:99:99:nobody:/home:/bin/false
```

yangdongsheng commented:
@LK4D4 I found my code does not affect "-" branches at all. And I tried it with
```
# ./bundles/1.10.0-dev/binary/docker cp fc85c95b2b82://etc/passwd -
passwd0100664000000000000000000000056312573124251010501 0ustar0000000000000000root:x:0:0:root:/root:/bin/sh
daemon:x:1:1:daemon:/usr/sbin:/bin/false
bin:x:2:2:bin:/bin:/bin/false
sys:x:3:3:sys:/dev:/bin/false
sync:x:4:100:sync:/bin:/bin/sync
mail:x:8:8:mail:/var/spool/mail:/bin/false
www-data:x:33:33:www-data:/var/www:/bin/false
operator:x:37:37:Operator:/var:/bin/false
ftp:x:83:83:ftp:/home/ftp:/bin/false
nobody:x:99:99:nobody:/home:/bin/false
```
Is this output expected?

BTW, I rebased my branch against master.

yangdongsheng commented:
@LK4D4 But I found an error in CI
```
ERROR: Build step failed with exception
java.lang.IllegalStateException: Invalid object ID 34 iota=35
```
Sorry I am not sure what that means. :(

yangdongsheng commented:
@LK4D4 Ha, I checked it again, the error I mentioned above disappeared, I guess there was something happened in CI server.

yangdongsheng commented:
@LK4D4  any more comment?

aaronlehmann commented:
This needs a rebase, but I recommend waiting for #18353, which makes changes to `ProgressReader`.

aaronlehmann commented on pkg/progressreader/progressreader.go:
```diff
@@ -11,15 +11,16 @@ import (
+	Current        int64
+	LastUpdate     int64
+	NewLines       bool
+	NewLineInClose bool
```
Instead of adding another option to `ProgressReader` that's specific to this use case, can you do something like this in `copyFromContainer`:

```go
defer func() {
        content.Close()
        // Format blank line
        cli.out.Write(sf.FormatStatus("", ""))
}()
```

thaJeztah commented:
ping @yangdongsheng can you address @aaronlehmann's comment https://github.com/docker/docker/pull/17014#discussion_r46639817? Also looks like it needs a rebase :cry:

yangdongsheng commented:
@thaJeztah I would like to wait for #18353 which is changing progressreader.

thaJeztah commented:
@yangdongsheng makes sense, thank you!

thaJeztah commented:
ping @yangdongsheng, https://github.com/docker/docker/pull/18353 was just merged. can you rebase this and update this for the changes in https://github.com/docker/docker/pull/18353? Thanks!

yangdongsheng commented:
@thaJeztah  Rebased, Please take a look. @aaronlehmann Good idea. I updated it. :)

thaJeztah commented:
Thanks @yangdongsheng!

ping @LK4D4 @aaronlehmann @cpuguy83 let's get this moving again! \o/

thaJeztah commented:
ping @LK4D4 @aaronlehmann @cpuguy83 could you have a look at this one? Thanks!

aaronlehmann commented on api/client/cp.go:
```diff
@@ -191,10 +193,29 @@ func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string, c
+		reader = progress.NewProgressReader(preArchive, progressOutput, size, "", "Copying")
+		defer func() {
+			// Format blank line
+			cli.out.Write(sf.FormatStatus("", ""))
```
Is this necessary, since you pass `newLines == true` to `NewProgressOutput`?

yangdongsheng replied:
The newLines to NewProgressOutput means the progress bar would be printed in a new line. But this def func here is going to append a '\n' at the end of progress bar. In other words, newLines == true means the '\n' at the start of progress bar, the "def func()" here means '\n' at the end of progress bar

aaronlehmann replied:
My understanding is that `newLines` causes the progress output to emit a new line when the progress reader is finished. It looks very similar to what you're doing here.

https://github.com/docker/docker/blob/master/pkg/streamformatter/streamformatter.go#L126
https://github.com/docker/docker/blob/master/pkg/progress/progressreader.go#L40

yangdongsheng replied:
Great!! Thanx @aaronlehmann , I don't like the ugly "def func" way in my patch too. I will investigate why I did not get what I want with NewLines == true. Anyway, I would take the use of the infrastructure you mentioned above. :)

aaronlehmann commented on api/types/types.go:
```diff
@@ -151,9 +151,12 @@ type CopyConfig struct {
 type ContainerPathStat struct {
 	Name       string      `json:"name"`
 	Size       int64       `json:"size"`
+	DSize      int64       `json:"dsize"`
```
Where is this set?

calavera replied:
it looks like it's not necessary in this struct, can we remove it?

yangdongsheng replied:
The dsize works when the path is a directory. size only report the dir size, but We want to get the Whole size of the files in this dir. Then I introduce a DSize here.

aaronlehmann commented on pkg/archive/copy.go:
```diff
@@ -151,11 +154,19 @@ func CopyInfoSourcePath(path string, followLink bool) (CopyInfo, error) {
 		IsDir:      stat.IsDir(),
 		RebaseName: rebaseName,
+		Size:       stat.Size(),
+		DSize:      dsize,
```
Both places that use this information ignore `Size` and use `DSize` instead if the path is a directory, so it seems like it would be simpler to only have one `Size` variable.

yangdongsheng replied:
But, I don't want to change the behavior of Size, which maybe introduce a regression. There is a possiblilty somebody want to get only the size of a dir path rather than the all files in it.

aaronlehmann replied:
I don't understand how it's possible to have a regression when the `Size` field didn't exist before this PR.

yangdongsheng replied:
But Size exists in "type ContainerPathStat struct" already, although we did not use Size here. And there are lots of other users of this struct. If the Size did not container its children before, I think that's possible to introduce regression.

aaronlehmann commented:
@LK4D4:

> @yangdongsheng This breaks - usage:
> [...]

At least with the current revision of the PR, it looks like the code path for "-" as a destination is not affected. I think you get that output because the behavior of `docker cp ... -` is to emit a tar stream on stdout.

Commit:
progressreader: fix the end of file checking.

When we are checking the end of file, we should not
assume it as read == 0. Because not all reader would
read a len==0 data out. This patch change it to check
the err == EOF.

Signed-off-by: Dongsheng Yang <yangds.fnst@cn.fujitsu.com>

yangdongsheng commented:
@aaronlehmann Thanx for your comment, I added a new commit to here. Please help to take a look :)

aaronlehmann commented on pkg/progress/progressreader.go:
```diff
@@ -37,7 +37,7 @@ func (p *Reader) Read(buf []byte) (n int, err error) {
 	}
 	if p.current-p.lastUpdate > updateEvery || err != nil {
-		p.updateProgress(err != nil && read == 0)
+		p.updateProgress(err == io.EOF)
```
How about just `err != nil`? We want the newline even in cases where the progress bar doesn't reach the end.

yangdongsheng replied:
okey, Fair point. Will update it soon Thanx

yangdongsheng commented:
@aaronlehmann Updated, PTAL. :)

aaronlehmann commented:
I still don't agree with adding a `DSize` field to `ContainerPathStat`, since this field is never used. Otherwise, this looks good.

yangdongsheng commented:
@aaronlehmann  Understandable, then let me try to check the all user of ContainerPathStat to see can we change the Size when path is dir. Thanx

aaronlehmann commented:
I would be satisfied if you just don't make any changes to the `ContainerPathStat` struct. It doesn't need a `DSize` field just because `CopyInfo` has it. That said, I'd be even happier if you added a single field to `CopyInfo` instead of two fields.

Commit:
archive: introduce Size in SrcInfo and get dir Size if Path is directory

Signed-off-by: Dongsheng Yang <yangds.fnst@cn.fujitsu.com>

Commit:
api/client/cp: provide progress indicator for docker cp

This commit make docker cp output the progress in copying,
something like docker pull.

Signed-off-by: Dongsheng Yang <yangds.fnst@cn.fujitsu.com>

yangdongsheng commented:
@aaronlehmann I changed the Size when Path is dir. Seems no regression here. Please take a look again. :)

aaronlehmann commented:
LGTM

thaJeztah commented:
Thanks @yangdongsheng, can you squash your commits? Let's see if we can get another "lgtm" and get this merged! \o/

tonistiigi commented:
Maybe this should be behind an option, as I may not want a performance hit of a full directory scan? If it is a live container this progress can also be very wrong as we don't pause while copying. When copying a directory the total is always bit different as `directory.Size` only the sums up file sizes and not the metadata info.

yangdongsheng commented:
Hi @tonistiigi , yes, that's good to make the tree walking optional. But currently, IIRC, the user of statPath() is only the cp command, and cp command have to walk the dir tree. So I believe there would not be any performance problem now.

About your idea, as there is no real user for this usecase, I am not sure how to do it better. Maybe someone need stat without Size. or maybe someone need Size but don't want to walk children. What about leave it here and change it when that's necessary? :)

aaronlehmann commented:
Personally, I think the directory scan is reasonable, because it will warm the OS' cache for metadata access during the copy. The main downside is that the acutal copying may be delayed if there are a lot of files to stat. I imagine that other file management tools that show progress information (for example, desktop GUIs) do a similar directory scan.

tonistiigi commented:
> I imagine that other file management tools that show progress information (for example, desktop GUIs) do a similar directory scan.

That's probably one of the reasons why cp/scp don't show this(for dirs).

Leaving out the performance, it is still the case that the measurement is never correct as stat uses only files but what is transferred is the tar package of these file(s). You can see it even with small files where after the download is completed the transfer size is bigger than 100%.

thaJeztah commented:
Based on the above concerns, I think we should close this PR; it's a "nice to have", and I'm not comfortable if it results in a degradation of performance, and possibly showing incorrect progress as well.

Thanks so much for your patience @yangdongsheng, I hope you understand

yangdongsheng commented:
@thaJeztah Fine, it's okey to me.:)

jeffabailey commented:
Why not have **any** indicator not necessarily progress?

I am currently copying a large file to a container and I have no idea if it's actually copying. I popped the container and I don't see a file there so I have to assume it's not.

Why do I have to assume that? Why can't the docker command tell me?

cpuguy83 commented:
Feel free to make an official proposal.
